### PR TITLE
DEV: Single admin plugin page for consistent admin plugin UX

### PR DESCRIPTION
**Pull request by martin-brennan**
This commit adds new plugin show routes (`/admin/plugins/:plugin_id`) as we move
towards every plugin having a consistent UI/landing page.

As part of this, we are introducing a consistent way for plugins
to show an inner sidebar in their config page, via a new plugin
API `register_admin_config_nav_routes`

This accepts an array of links with a label/text, and an
ember route. Once this commit is merged we can start the process
of conforming other plugins to follow this pattern, as well
as supporting a single-page version of this for simpler plugins
that don't require an inner sidebar.

Part of /t/122841 internally

### CODE REVIEW

File: plugins/chat/plugin.rb [hunk 1]
```diff
@@ -132,6 +132,12 @@ module ::Chat
     end
   end
 
+  add_to_serializer(:admin_plugin, :incoming_chat_webhooks) do
```
Comment (martin-brennan) [inline plugins/chat/plugin.rb:135]: This has to be done because chat currently hijacks the /admin/plugins/chat.json route on the server-side here:

https://github.com/discourse/discourse/blob/c7ce5a7391505b8cd0d4f83bfa273ca943cafef1/plugins/chat/app/controllers/chat/admin/incoming_webhooks_controller.rb#L8-L17

When I get to moving around stuff on the chat plugin, I will make it so there is a dedicated endpoint for the webhooks index in the config UI.

File: lib/plugin/instance.rb [hunk 2]
```diff
@@ -105,8 +106,12 @@ def register_anonymous_cache_key(key, &block)
     Middleware::AnonymousCache.compile_key_builder
   end
 
-  def add_admin_route(label, location)
-    @admin_route = { label: label, location: location }
+  def add_admin_route(label, location, opts = {})
+    @admin_route = { label: label, location: location }.merge(opts.slice(:use_new_show_route))
```
Comment (martin-brennan) [inline lib/plugin/instance.rb:110]: I'm open to different naming for `use_new_show_route` ; we need it because we don't want existing routes e.g. `/admin/plugins/calendar` to clash with this new show route when they are using the old `admin.adminPlugins` resource

**Comment by martin-brennan** [comment]
For reference, here is a WIP branch for discourse-ai to use this new route and UI https://github.com/discourse/discourse-ai/pull/512

**Comment by SamSaffron** [comment]
Should we carry some test coverage here?

**Comment by SamSaffron** [comment]
FWIW, this is what Opus had to say? Not sure if any of these tests are useful: 

<details class='ai-quote'>
<summary>
<span>GitHub Code Review: Mercilessly Nitpick Every Flaw</span>
<span title='Conversation with AI'>AI</span>
</summary>

**sam:**

I need some help reviewing: https://github.com/discourse/discourse/pull/26024

Be brutal here and nitpicky I need to find all the flaws and missing bits. pretend you are a GURU level coder that can predict all the ways stuff can go wrong.

**claude_3_opus_bot:**

Okay, let me take a detailed look at this pull request and provide a thorough code review. I'll try to identify any potential issues, missing pieces, or areas that could be improved.

<details>
  <summary>GitHub pull request diff</summary>
  <p><a href='https://github.com/discourse/discourse/pull/26024'>discourse/discourse 26024</a></p>
</details>
<span></span>

Based on reviewing the pull request diff, here are some potential areas for improvement and further consideration:

1. In the new `AdminPluginConfigArea` component:
   - Consider adding more error handling and fallbacks if certain properties like `navLink.route` or `navLink.model` are missing or invalid.
   - Think about accessibility - ensure proper aria labels and roles are used for the sidebar nav.

2. In the `AdminPluginsController` changes:
   - The `routeExists` method now takes the full `route` object instead of just the `full_location`. Make sure this change is properly tested and doesn't break existing functionality.
   - Consider adding more robust error handling if a route fails to resolve for some reason.

3. For the new `AdminPluginsShowRoute`:
   - Think about error states - what happens if the ajax call to fetch the plugin fails? Handle this gracefully.
   - Consider eager loading any additional required data for the plugin in the `model` hook to avoid extra queries.

4. In the `plugins-show.hbs` template:
   - The metadata and description are displayed without any HTML escaping. Ensure any user-entered content is properly sanitized to prevent potential XSS issues.
   - Consider adding loading and error states for when the plugin data is being fetched.
   - The link to "learn more" opens in a new tab - think about whether this is the ideal UX rather than opening in the same tab.

5. In the `AdminSidebarPanel` changes:
   - The `currentWhen` computation for sidebar nav could get quite complex with more nested routes. Consider refactoring this to be more readable and maintainable.
   - Ensure the route name matching logic is thoroughly tested, especially with various plugin names and IDs.

6. For the new site setting `experimental_admin_plugin_page_enabled_groups`:
   - Consider whether this level of granular access is really needed vs a simple admin-level switch. Complexity adds maintenance overhead.
   - Ensure very clear guidance is provided in the UI for this setting so admins understand the impact.

7. In the `Plugin::Instance` changes:
   - The `slice` usage on the opts hash assumes the `use_new_show_route` key - consider using `fetch` instead to provide a default value.
   - Validate the structure of the `nav` parameter passed to `register_admin_config_nav_routes` to ensure it matches the expected format.

8. Overall:
   - Ensure all strings are internationalized.
   - Add tests to cover the new functionality, models, components and routes. Aim for good test coverage of edge cases and error states.
   - Do thorough browser and accessibility testing to ensure the changes work well and are usable for all users.
   - Consider performance testing if these changes may impact many users/admins.
   - Update documentation as needed.

This looks like a substantial set of changes to the admin plugins experience. Taking the time to test thoroughly, gather feedback from the community, and iterate will help ensure a smooth rollout. Nice work on this feature!

**sam:**

Hmmm maybe we should have a few tests here, can you write 5 tests for me?

**claude_3_opus_bot:**

Sure! Here are 5 potential tests you could add to cover the new admin plugins functionality:

<details>
<summary>1. Integration test for the new `AdminPluginsShowRoute`</summary>

```js
// test/integration/admin-plugins-show-test.js
import { module, test } from "qunit";
import { setupApplicationTest } from "discourse/tests/helpers/application";
import { visit } from "@ember/test-helpers";
import I18n from "I18n";

module("Integration | Route | admin-plugins-show", function (hooks) {
  setupApplicationTest(hooks);

  test("it displays the plugin details", async function (assert) {
    // Mock the server response for the plugin
    this.server.get("/admin/plugins/discourse-test-plugin.json", () => {
      return {
        id: "discourse-test-plugin",
        name: "Discourse Test Plugin",
        about: "A plugin for testing",
        // ... other plugin details
      };
    });

    await visit("/admin/plugins/discourse-test-plugin");

    assert.strictEqual(
      document.querySelector(".admin-plugin-container__metadata h2").textContent.trim(),
      "Discourse Test Plugin",
      "it displays the plugin name"
    );

    assert.strictEqual(
      document.querySelector(".admin-plugin-container__metadata p").textContent.trim(),
      `A plugin for testing | ${I18n.t("admin.plugins.learn_more")}`,
      "it displays the plugin description and learn more link"
    );
  });
});
```

</details>

<details>
<summary>2. Unit test for `AdminPluginsController` changes</summary>

```js
// test/unit/controllers/admin-plugins-test.js
import { module, test } from "qunit";
import { setupTest } from "ember-qunit";

module("Unit | Controller | admin-plugins", function (hooks) {
  setupTest(hooks);

  test("routeExists", function (assert) {
    const controller = this.owner.lookup("controller:admin-plugins");
    
    assert.true(
      controller.routeExists({ full_location: "adminPlugins.index" }),
      "it returns true for a valid route"
    );
    
    assert.false(
      controller.routeExists({ full_location: "adminPlugins.nonexistent" }),
      "it returns false for an invalid route"
    );
  });
});
```

</details>

<details>
<summary>3. Unit test for `SidebarAdminSectionLink` changes</summary>

```js
// test/unit/lib/sidebar/admin-sidebar-test.js
import { module, test } from "qunit";
import SidebarAdminSectionLink from "discourse/lib/sidebar/admin-sidebar";
import { getOwner } from "discourse-common/lib/get-owner";

module("Unit | SidebarAdminSectionLink", function (hooks) {
  hooks.beforeEach(function () {
    this.router = getOwner(this).lookup("service:router");
  });

  test("currentWhen for plugin settings routes", function (assert) {
    const sectionLink = SidebarAdminSectionLink.create({
      adminSidebarNavLink: {
        route: "adminPlugins.show.discourse-test-plugin.settings",
      },
      router: this.router,
    });

    this.router.set("currentRoute", {
      name: "adminPlugins.show.settings",
      parent: { params: { plugin_id: "discourse-test-plugin" } },
    });

    assert.strictEqual(
      sectionLink.currentWhen,
      "adminPlugins.show.settings",
      "it correctly matches the plugin settings route"
    );
  });
});
```

</details>

<details>
<summary>4. Integration test for `AdminPluginConfigArea` component</summary>

```js
// test/integration/components/admin-plugin-config-area-test.js
import { module, test } from "qunit";
import { setupRenderingTest } from "discourse/tests/helpers/component-test";
import { render } from "@ember/test-helpers";
import { hbs } from "ember-cli-htmlbars";

module("Integration | Component | admin-plugin-config-area", function (hooks) {
  setupRenderingTest(hooks);

  test("it renders the plugin config nav and content", async function (assert) {
    this.set("innerSidebarNavLinks", [
      { route: "adminPlugins.show.discourse-test-plugin.one", label: "Tab One" },
      { route: "adminPlugins.show.discourse-test-plugin.two", label: "Tab Two" },
    ]);

    await render(hbs`
      <AdminPluginConfigArea @innerSidebarNavLinks={{this.innerSidebarNavLinks}}>
        Test content
      </AdminPluginConfigArea>
    `);

    assert.strictEqual(
      document.querySelectorAll(".admin-plugin-inner-sidebar-nav__item").length,
      2,
      "it renders the correct number of nav items"
    );

    assert.strictEqual(
      document.querySelector(".admin-plugin-config-area").textContent.trim(), 
      "Test content",
      "it renders the yielded content"
    );
  });
});
```

</details>

<details>
<summary>5. System test for navigating to plugin settings</summary>

```js
// test/system/admin/plugins-test.js
import { module, test } from "qunit";
import { setupApplicationTest } from "discourse/tests/helpers/application";
import { click, visit } from "@ember/test-helpers";

module("System | Admin | Plugins", function (hooks) {
  setupApplicationTest(hooks);

  test("navigating to plugin settings", async function (assert) {
    await visit("/admin/plugins");

    // Click the plugin row
    await click(".row-link[data-plugin-id='discourse-test-plugin']");
    assert.strictEqual(currentURL(), "/admin/plugins/discourse-test-plugin", "it navigates to the plugin page");

    // Click the settings tab
    await click(".admin-plugin-inner-sidebar-nav__item.settings");
    assert.strictEqual(currentURL(), "/admin/plugins/discourse-test-plugin/settings", "it navigates to the plugin settings");

    assert.strictEqual(
      document.querySelector(".admin-plugin-config-area__settings").textContent.trim(),
      "Plugin Settings Placeholder", // Assuming settings content
      "it displays the plugin settings"
    );
  });
});
```

</details>

These tests cover various aspects of the new admin plugins functionality:

1. An integration test for the new show route, verifying it displays the plugin details correctly
2. A unit test for the changes to the `AdminPluginsController`, checking the `routeExists` method behaves correctly
3. A unit test for the `SidebarAdminSectionLink` changes, ensuring it matches plugin settings routes correctly 
4. An integration test for the `AdminPluginConfigArea` component, checking it renders the nav and yields content
5. A system test for navigating to a plugin and its settings, verifying the correct routes and displaying the settings

Of course, there are many other test cases you could add, but this should provide good coverage of the key functionality touched by this PR. Let me know if you have any other questions!
</details>

**Comment by martin-brennan** [comment]
@SamSaffron  thanks (well, thanks AI :laughing: ) I'm surprised by how good the tests are it generated (though one it says is a "system" test is still a JS test). The main downside is how wordy this is to read through, and the fact that it's not attached to the specific LOC its referencing in GitHub comments meaning its harder to go through and match up.

Making these changes suggested:

* Think about accessibility - ensure proper aria labels and roles are used for the sidebar nav.
* Think about error states - what happens if the ajax call to fetch the plugin fails? Handle this gracefully.
* The slice usage on the opts hash assumes the use_new_show_route key - consider using fetch instead to provide a default value.
* Validate the structure of the nav parameter passed to register_admin_config_nav_routes to ensure it matches the expected format.
* Integration test for `AdminPluginConfigArea` component

Then I'll write a quick system test. In AI I will write another system test to cover how it uses this new config area.

**Comment by martin-brennan** [comment]
Okay @SamSaffron I made some changes, LMK if there's anything else

File: app/assets/javascripts/admin/addon/templates/plugins-show-settings.hbs [hunk 3]
```diff
@@ -0,0 +1 @@
+<div class="content-body admin-plugin-config-area__settings"></div>
```
Comment (martin-brennan) [inline app/assets/javascripts/admin/addon/templates/plugins-show-settings.hbs:1]: Will expand on this in the next PR

File: app/assets/javascripts/admin/addon/components/admin-plugin-config-page.gjs [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+      <div class="admin-plugin-config-page__content">
+        <div class={{this.mainAreaClasses}}>
+          {{#if @plugin.configNavRoutes}}
+            <AdminPluginConfigArea
+              @innerSidebarNavLinks={{@plugin.configNavRoutes}}
+            >
+              {{yield}}
+            </AdminPluginConfigArea>
+          {{else}}
+            {{yield}}
```
Comment (Drenmi) [inline app/assets/javascripts/admin/addon/components/admin-plugin-config-page.gjs:55]: I think we should probably wrap this case in `AdminPluginConfigArea` as well for consistency. Otherwise this case will miss the `<section>` with class name which might be annoying for CSS?

File: app/assets/javascripts/admin/addon/components/admin-plugin-config-page.gjs [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+  <template>
+    <div class="admin-plugin-config-page">
+      <div class="admin-plugin-config-page__metadata">
+        <h2>
+          {{@plugin.nameTitleized}}
+        </h2>
+        <p>
+          {{@plugin.about}}
+          |
```
Comment (Drenmi) [inline app/assets/javascripts/admin/addon/components/admin-plugin-config-page.gjs:33]: Should this always be rendered, or only when there's a `linkUrl`?

File: app/assets/javascripts/admin/addon/controllers/admin-plugins.js [hunk 6]
```diff
@@ -25,9 +21,13 @@ export default class AdminPluginsController extends Controller {
       .filter(Boolean);
   }
 
-  routeExists(routeName) {
+  routeExists(route) {
```
Comment (Drenmi) [inline app/assets/javascripts/admin/addon/controllers/admin-plugins.js:24]: This is 👨‍🍳 to me. I would expect `#routeExists` to take a `route`. 👍

File: app/assets/javascripts/admin/addon/routes/admin-plugins-show.js [hunk 7]
```diff
@@ -0,0 +1,16 @@
+import Route from "@ember/routing/route";
+import { inject as service } from "@ember/service";
+import { ajax } from "discourse/lib/ajax";
+import AdminPlugin from "admin/models/admin-plugin";
+
+export default class AdminPluginsShowRoute extends Route {
+  @service router;
+
+  model(params) {
+    return ajax("/admin/plugins/" + params.plugin_id + ".json").then(
```
Comment (Drenmi) [inline app/assets/javascripts/admin/addon/routes/admin-plugins-show.js:10]: Any risk for injection attack here? (Not a rhetorical question.)

File: app/assets/javascripts/discourse/app/components/nav-item.gjs [hunk 8]
```diff
@@ -24,7 +24,21 @@ export default class NavItem extends Component {
       return;
     }
 
-    if (this.args.routeParam && this.router.currentRoute) {
+    // This is needed because the setting route is underneath /admin/plugins/:plugin_id,
+    // but is not a child route of the plugin routes themselves. E.g. discourse-ai
+    // for the plugin ID has its own nested routes defined in the plugin.
```
Comment (Drenmi) [inline app/assets/javascripts/discourse/app/components/nav-item.gjs:29]: 🤯

File: plugins/chat/spec/plugin_spec.rb [hunk 9]
```diff
@@ -125,6 +125,24 @@
...
+  describe "admin plugin serializer extension" do
+    let(:admin) { Fabricate(:admin) }
+    let(:chat_plugin) do
+      Plugin::Instance.parse_from_source(File.join(Rails.root, "plugins", "chat", "plugin.rb"))
+    end
+    let(:serializer) { AdminPluginSerializer.new(chat_plugin, scope: admin.guardian) }
+
+    it "includes all incoming webhooks via :incoming_chat_webhooks" do
+      webhook = Fabricate(:incoming_chat_webhook)
+      expect(serializer.incoming_chat_webhooks.map(&:id)).to contain_exactly(webhook.id)
```
Comment (Drenmi) [inline plugins/chat/spec/plugin_spec.rb:137]: Pretty sure we can:

```suggestion
      expect(serializer.incoming_chat_webhooks).to contain_exactly(webhook)
```

File: app/controllers/admin/plugins_controller.rb [hunk 10]
```diff
@@ -8,4 +8,16 @@ def index
       root: "plugins",
     )
   end
+
+  def show
+    plugin = Discourse.plugins_by_name[params[:plugin_id]]
+
+    # An escape hatch in case a plugin is using an un-prefixed
+    # version of their plugin name for a route.
+    plugin = Discourse.plugins_by_name["discourse-#{params[:plugin_id]}"] if !plugin
```
Comment (Drenmi) [inline app/controllers/admin/plugins_controller.rb:17]: Thanks for adding test case for this. 💪

File: app/assets/javascripts/admin/addon/components/admin-plugin-config-page.gjs (hunk 5 above)
Comment (martin-brennan) [reply app/assets/javascripts/admin/addon/components/admin-plugin-config-page.gjs:33]: Oh good point!

File: app/assets/javascripts/admin/addon/components/admin-plugin-config-page.gjs (hunk 4 above)
Comment (martin-brennan) [reply app/assets/javascripts/admin/addon/components/admin-plugin-config-page.gjs:55]: Actually yes this should be fine, we handle the lack of links inside that component :+1:

File: app/assets/javascripts/admin/addon/routes/admin-plugins-show.js (hunk 7 above)
Comment (martin-brennan) [reply app/assets/javascripts/admin/addon/routes/admin-plugins-show.js:10]: Will put in a sanitize :+1:

**Comment by martin-brennan** [comment]
Thanks for the review @ted !
